source-batchsql/main.go

```diff
@@ -0,0 +1,480 @@
+	// TODO(wgd): Add network tunnel support
+}
+
+type advancedConfig struct {
+	SSLMode string `json:"sslmode,omitempty" jsonschema:"title=SSL Mode,description=Overrides SSL connection behavior by setting the 'sslmode' parameter.,enum=disable,enum=allow,enum=prefer,enum=require,enum=verify-ca,enum=verify-full"`
+}
+
+// Validate checks that the configuration possesses all required properties.
+func (c *Config) Validate() error {
+	var requiredProperties = [][]string{
+		{"address", c.Address},
+		{"user", c.User},
+		{"password", c.Password},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	return nil
+}
+
+// SetDefaults fills in the default values for unset optional parameters.
+func (c *Config) SetDefaults() {
+	// The address config property should accept a host or host:port
+	// value, and if the port is unspecified it should be the PostgreSQL
+	// default 5432.
+	if !strings.Contains(c.Address, ":") {
+		c.Address += ":5432"
+	}
+}
+
+// ToURI converts the Config to a DSN string.
+func (c *Config) ToURI() string {
+	var address = c.Address
+	var uri = url.URL{
+		Scheme: "postgres",
+		Host:   address,
+		User:   url.UserPassword(c.User, c.Password),
+	}
+	if c.Database != "" {
+		uri.Path = "/" + c.Database
+	}
+	var params = make(url.Values)
+	if c.Advanced.SSLMode != "" {
+		params.Set("sslmode", c.Advanced.SSLMode)
+	}
+	if len(params) > 0 {
+		uri.RawQuery = params.Encode()
+	}
+	return uri.String()
+}
+
+// Resource represents the capture configuration of a single resource binding.
+type Resource struct {
+	Name         string   `json:"name" jsonschema:"title=Name,description=The unique name of this resource."`
+	Template     string   `json:"template" jsonschema:"title=Query Template,description=The query template (pkg.go.dev/text/template) to execute to fetch new data,multiline=true"`
+	Cursor       []string `json:"cursor" jsonschema:"title=Cursor Columns,description=The names of columns which should be persisted between query executions as a cursor"`
+	PollInterval string   `json:"poll,omitempty" jsonschema:"title=Poll Interval,description=How often to execute the fetch query."`
+}
+
+// Validate checks that the resource spec possesses all required properties.
+func (r Resource) Validate() error {
+	var requiredProperties = [][]string{
+		{"name", r.Name},
+		{"template", r.Template},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	if _, err := template.New("query").Parse(r.Template); err != nil {
+		return fmt.Errorf("error parsing template: %w", err)
+	}
+	if r.PollInterval != "" {
+		if _, err := time.ParseDuration(r.PollInterval); err != nil {
+			return fmt.Errorf("invalid poll interval %q: %w", r.PollInterval, err)
+		}
+	}
+	return nil
+}
+
+// documentMetadata contains the source metadata located at /_meta
+type documentMetadata struct {
```

Comment:
For now I've chosen to have the document metadata be as simple as possible, with a "polled at" timestamp and a "result row index" counter which I see as being kind of analagous to how a file source has file and offset metadata.